docs/rfc/rfc-003-performance-questions.md

```diff
@@ -0,0 +1,193 @@
+The Cosmos-SDK runs the ABCI application within the same process as Tendermint.
+When an application is run in the same process as Tendermint, a serialization penalty
+is not paid. This is because the local ABCI client does not serialize method calls
+and instead passes the protobuf type through through directly.
+
+Serialization and deserialization in the gRPC and socket protocol ABCI methods
+may cause slowdown. While these may cause issue, they are not part of the primary
+usecase of Tendermint and do not necessarily need to be addressed at this time.
+
+### RPC
+
+#### The Query API
+
+* Not yet clear what the use case is for this or if this is the best solve for the need for users to query TM state? 
+* Should we solve this within Tendermint?
+
+#### RPC Serialization
+
+The Tendermint RPC uses a modified version of JSON-RPC. This RPC powers the `broadcast_tx_*` methods,
+which is a critical method for adding transactions to Tendermint at the moment. This method is
+likely invoked quite frequently on popular networks. Being able to perform efficiently
+on this common and critical operation is very important. The current JSON-RPC implementation
+relies heavily on type introspection via reflection, which is known to be very slow in
+Go. We should therefore produce benchmarks of this method to determine how much overhead
+we are added to what is likely to be a very common operation.
+
+The other JSON-RPC methods are much less critical to the core functionality of Tendermint.
+While there may other points of performance consideration within the RPC, methods that do not
+receive high volumes of requests should not be prioritized for performance consideration.
+
+NOTE: There is ongoing work to inspect and alter the JSON-RPC framework, so much of these
+RPC-related performance considerations can either wait until that work is done or be
+considered concordantly with the in-flight changes to the JSON-RPC.
+
+### Protocol
+
+#### Gossiping messages
+
+Currently, for any validator to successfully vote in a consensus _step_, it must
+receive votes from greater than 2/3 of the validators on the network. In many cases,
+it's preferable to receive as many votes as possible from correct validators.
+
+This produces a quadratic increase in messages that are communicated as more validators join the network.
+(Each of the N validators must communicate with all other N-1 validators).
+
+This large number of messages communicated per step has been identified to impact
+performance of the protocol. Given this that number of messages communicated has been
+identified as a bottleneck, it would be extremely valuable to gather data on how long
+it takes for popular chains with many validators to gather all votes within a step.
+
+#### Tx hashes
+
+Using a faster hash algorithm for Tx hashes is currently a point of discussion
+in Tendermint. It is currently unknown if hashing transactions in the Mempool forms
```

Comment:
I'm pretty sure this is not a real issue. A tx hash is peanuts compared to signature verification.

Reply:
I think it makes sense to discuss/callout signature aggregation and similar kinds of proposals somewhere in this document as it provides a method for reducing overall message size, and also a space to talk about sig-verification in general.

Reply:
It's quite likely to not be a serious source of performance issue, but it is currently cited as potential motivation for the [modular hasher change proposal](https://github.com/tendermint/tendermint/pull/6773/files#diff-06a88f656cef810abba7234cd95440bebeb9a9899c8dc1264b99c930c0f26c48R13). 

From the [references on that proposal](https://github.com/tendermint/tendermint/pull/6773/files#diff-06a88f656cef810abba7234cd95440bebeb9a9899c8dc1264b99c930c0f26c48R137-R142), it appears that concerns about the speed of hashing has surfaced before. While it may in fact not be an issue, it's worth documenting and understanding its performance so that we can more productively discuss tradeoffs of different hash algorithms in the future.

A goal of mine in writing this RFC is to gather points where Tendermint is thought to be slow so that we can sort out some relative ordering of these concerns. This will hopefully be helpful going forward in allowing us to accurately of  focus our efforts on the most pressing areas of performance concern. Certain areas of Tendermint may be slower-than-an-optimal-solution such as hashing, but likely negligible in the grand scheme of things.

I'll add a section on digital signatures, since that's definitely worth discussing more in this document!